cpp/splinepy/utils/arrays.hpp

```diff
@@ -0,0 +1,283 @@
+#pragma once
+
+#include <array>
+#include <cmath>
+#include <algorithm>
+#include <numeric>
+
+namespace splinepy::utils {
+
+using namespace splinelib::sources;  
+
+/// Elementwise subtraction
+template<typename T, typename NamedType, std::size_t dim>
+inline void FirstMinusSecondEqualsThird(
+    const std::array<NamedType, dim>& first,
+    const T* second, /* c array */
+    std::array<T, dim>& third) {
+  for (size_t i{}; i < dim; ++i) {
+    // following line should raise error during compile time
+    // if T != NameType::Type_
+    third[i] = static_cast<T>(first[i]) - second[i];
+  }
+}
+
+/// Elementwise mean
+template<typename T, std::size_t dim>
+inline std::array<T, dim> Mean(const std::array<T, dim>& arr1,
+                               const std::array<T, dim>& arr2) {
+  std::array<T, dim> out;
+  for (int i{0}; i < dim; ++i) {
+    out[i] = (arr1[i] + arr2[i]) * .5;
+  }
+  return out;
+}
+
+/// Elementwise mean overload, you can specified return value_type.
+/// Useful for returning NamedType.
+template<typename ReturnT, typename T, std::size_t dim>
+inline std::array<ReturnT, dim> Mean(const std::array<T, dim>& arr1,
+                                     const std::array<T, dim>& arr2) {
+  std::array<ReturnT, dim> out;
+  for (int i{0}; i < dim; ++i) {
+    out[i] = ReturnT{(arr1[i] + arr2[i]) * .5};
+  }
+  return out;
+}
+
+/// Dot product
+template<typename T1, typename T2, std::size_t dim>
+inline T1 Dot(const std::array<T1, dim>& arr1,
+              const std::array<T2, dim>& arr2) {
+  T1 dotted{}; /* default value should be 0. */
+  for (int i{0}; i < dim; ++i) {
+    dotted += arr1[i] * static_cast<T1>(arr2[i]);
+  }
+  return dotted;
+}
+
+/// AAt
```

Comment:
Please extend the description, e.g., write a short equation in the style `r_ij = a_ik * b_jk`